src/lib/parameters/param.h

```diff
@@ -467,18 +467,20 @@ __END_DECLS
 // param is a C-interface. This means there is no overloading, and thus no type-safety for param_get().
 // So for C++ code we redefine param_get() to inlined overloaded versions, which gives us type-safety
 // w/o having to use a different interface
-static inline int param_get(param_t param, float *val)
+static inline int param_get_cplusplus(param_t param, float *val)
```

Comment:
This breaks linking on **gcc-linaro-7.5.0-2019.12-x86_64_arm-linux-gnueabihf**

Reply:
Please provide the error.

Reply:
```
[691/691] Linking CXX executable bin/px4
FAILED: bin/px4 
: && /home/jake/hexagon/Qualcomm/ARM_Tools/gcc-linaro-7.5.0-2019.12-x86_64_arm-linux-gnueabihf/bin/arm-linux-gnueabihf-g++ --sysroot=/home/jake/hexagon/Qualcomm/ARM_Tools/gcc-linaro-7.5.0-2019.12-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/libc  -O2 -g -DNDEBUG  -fuse-ld=gold   -Wl,--export-dynamic -rdynamic platforms/posix/CMakeFiles/px4.dir/px4muorb_stub.c.o platforms/posix/CMakeFiles/px4.dir/src/px4/common/main.cpp.o platforms/posix/CMakeFiles/px4.dir/apps.cpp.o platforms/posix/CMakeFiles/px4.dir/home/jake/hexagon/Qualcomm/Hexagon_SDK/3.1/libs/common/rpcmem/src/rpcmem.c.o  -o bin/px4 -L/home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/external/Install/lib -Wl,-rpath,/home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/external/Install/lib -Wl,--start-group platforms/posix/src/px4/common/libpx4_layer.a src/lib/parameters/libparameters.a src/modules/uORB/libmodules__uORB.a src/drivers/batt_smbus/libdrivers__batt_smbus.a src/drivers/camera_trigger/libdrivers__camera_trigger.a src/drivers/differential_pressure/ets/libdrivers__ets_airspeed.a src/drivers/differential_pressure/ms4525/libdrivers__ms4525_airspeed.a src/drivers/differential_pressure/ms5525/libdrivers__ms5525_airspeed.a src/drivers/differential_pressure/sdp3x/libdrivers__sdp3x_airspeed.a src/drivers/distance_sensor/cm8jl65/libdrivers__cm8jl65.a src/drivers/distance_sensor/leddar_one/libdrivers__distance_sensor__leddar_one.a src/drivers/distance_sensor/ll40ls/libdrivers__ll40ls.a src/drivers/distance_sensor/ll40ls_pwm/libdrivers__ll40ls_pwm.a src/drivers/distance_sensor/mappydot/libdrivers__mappydot.a src/drivers/distance_sensor/mb12xx/libdrivers__mb12xx.a src/drivers/distance_sensor/pga460/libdrivers__pga460.a src/drivers/distance_sensor/sf0x/libdrivers__sf0x.a src/drivers/distance_sensor/sf1xx/libdrivers__sf1xx.a src/drivers/distance_sensor/srf02/libdrivers__distance_sensor__srf02.a src/drivers/distance_sensor/teraranger/libdrivers__distance_sensor__teraranger.a src/drivers/distance_sensor/tfmini/libdrivers__tfmini.a src/drivers/distance_sensor/ulanding_radar/libdrivers__distance_sensor__ulanding_radar.a src/drivers/distance_sensor/vl53l0x/libdrivers__distance_sensor__vl53l0x.a src/drivers/gps/libdrivers__gps.a src/drivers/pwm_out_sim/libdrivers__pwm_out_sim.a src/drivers/qshell/posix/libdrivers__qshell__posix.a src/drivers/rc_input/libdrivers__rc_input.a src/modules/airspeed_selector/libmodules__airspeed_selector.a src/modules/attitude_estimator_q/libmodules__attitude_estimator_q.a src/modules/camera_feedback/libmodules__camera_feedback.a src/modules/commander/libmodules__commander.a src/modules/dataman/libmodules__dataman.a src/modules/ekf2/libmodules__ekf2.a src/modules/events/libmodules__events.a src/modules/fw_att_control/libmodules__fw_att_control.a src/modules/fw_pos_control_l1/libmodules__fw_pos_control_l1.a src/modules/land_detector/libmodules__land_detector.a src/modules/landing_target_estimator/libmodules__landing_target_estimator.a src/modules/local_position_estimator/libmodules__local_position_estimator.a src/modules/logger/libmodules__logger.a src/modules/mavlink/libmodules__mavlink.a src/modules/mc_att_control/libmodules__mc_att_control.a src/modules/mc_hover_thrust_estimator/libmodules__mc_hover_thrust_estimator.a src/modules/mc_pos_control/libmodules__mc_pos_control.a src/modules/mc_rate_control/libmodules__mc_rate_control.a src/modules/muorb/krait/libmodules__muorb__krait.a src/modules/muorb/test/libmodules__muorb__test__muorb_test.a src/modules/navigator/libmodules__navigator.a src/modules/rc_update/libmodules__rc_update.a src/modules/rover_pos_control/libmodules__rover_pos_control.a src/modules/sensors/libmodules__sensors.a src/modules/simulator/libmodules__simulator.a src/modules/simulator/battery_simulator/libmodules__simulator__battery_simulator.a src/modules/vmount/libdrivers__vmount.a src/modules/vtol_att_control/libmodules__vtol_att_control.a src/systemcmds/esc_calib/libsystemcmds__esc_calib.a src/systemcmds/led_control/libsystemcmds__led_control.a src/systemcmds/mixer/libsystemcmds__mixer.a src/systemcmds/motor_ramp/libsystemcmds__motor_ramp.a src/systemcmds/motor_test/libsystemcmds__motor_test.a src/systemcmds/param/libsystemcmds__param.a src/systemcmds/perf/libsystemcmds__perf.a src/systemcmds/pwm/libsystemcmds__pwm.a src/systemcmds/sd_bench/libsystemcmds__sd_bench.a src/systemcmds/shutdown/libsystemcmds__shutdown.a src/systemcmds/topic_listener/libsystemcmds__topic_listener.a src/systemcmds/tune_control/libsystemcmds__tune_control.a src/systemcmds/ver/libsystemcmds__ver.a src/systemcmds/work_queue/libsystemcmds__work_queue.a -L/home/jake/hexagon/Qualcomm/Hexagon_SDK/3.1/libs/common/remote/ship/UbuntuARM_Debug -lsdsprpc -lpthread -lm -lrt -Wl,--end-group -L/home/jake/hexagon/Qualcomm/Hexagon_SDK/3.1/libs/common/remote/ship/UbuntuARM_Debug -lsdsprpc src/lib/drivers/smbus/libdrivers__smbus.a src/lib/drivers/airspeed/libdrivers__airspeed.a src/lib/drivers/rangefinder/libdrivers_rangefinder.a src/lib/mixer_module/libmixer_module.a src/lib/output_limit/liboutput_limit.a src/lib/rc/librc.a src/lib/airspeed_validator/libAirspeedValidator.a src/lib/ecl/airdata/libecl_airdata.a src/modules/commander/failure_detector/libfailure_detector.a src/modules/commander/Arming/PreFlightCheck/libPreFlightCheck.a src/modules/commander/Arming/ArmAuthorization/libArmAuthorization.a src/modules/commander/Arming/HealthFlags/libHealthFlags.a src/lib/ecl/EKF/libecl_EKF.a src/lib/ecl/geo_lookup/libecl_geo_lookup.a src/modules/ekf2/Utility/libEkf2Utility.a src/modules/fw_pos_control_l1/launchdetection/liblaunchdetection.a src/modules/fw_pos_control_l1/runway_takeoff/librunway_takeoff.a src/lib/tecs/libtecs.a src/modules/mc_att_control/AttitudeControl/libAttitudeControl.a src/modules/mc_hover_thrust_estimator/libzero_order_hover_thrust_ekf.a src/lib/controllib/libcontrollib.a src/modules/mc_pos_control/PositionControl/libPositionControl.a src/lib/flight_tasks/libFlightTasks.a src/lib/flight_tasks/tasks/ManualAltitudeSmoothVel/libFlightTaskManualAltitudeSmoothVel.a src/lib/flight_tasks/tasks/ManualPositionSmooth/libFlightTaskManualPositionSmooth.a src/lib/flight_tasks/tasks/ManualPositionSmoothVel/libFlightTaskManualPositionSmoothVel.a src/lib/flight_tasks/tasks/ManualPosition/libFlightTaskManualPosition.a src/lib/flight_tasks/tasks/AutoLineSmoothVel/libFlightTaskAutoLineSmoothVel.a src/lib/flight_tasks/tasks/AutoMapper/libFlightTaskAutoMapper.a src/lib/flight_tasks/tasks/AutoFollowMe/libFlightTaskAutoFollowMe.a src/lib/flight_tasks/tasks/Auto/libFlightTaskAuto.a src/lib/avoidance/libavoidance.a src/lib/flight_tasks/tasks/Offboard/libFlightTaskOffboard.a src/lib/flight_tasks/tasks/Failsafe/libFlightTaskFailsafe.a src/lib/flight_tasks/tasks/Descend/libFlightTaskDescend.a src/lib/flight_tasks/tasks/Transition/libFlightTaskTransition.a src/lib/flight_tasks/tasks/Orbit/libFlightTaskOrbit.a src/lib/flight_tasks/tasks/ManualAltitudeSmooth/libFlightTaskManualAltitudeSmooth.a src/lib/flight_tasks/tasks/ManualAltitude/libFlightTaskManualAltitude.a src/lib/flight_tasks/tasks/Manual/libFlightTaskManual.a src/lib/flight_tasks/tasks/Utility/libFlightTaskUtility.a src/lib/flight_tasks/tasks/FlightTask/libFlightTask.a src/lib/bezier/libbezier.a src/lib/weather_vane/libWeatherVane.a src/lib/collision_prevention/libCollisionPrevention.a src/modules/mc_pos_control/Takeoff/libTakeoff.a src/lib/hysteresis/libhysteresis.a src/lib/circuit_breaker/libcircuit_breaker.a src/modules/mc_rate_control/RateControl/libRateControl.a src/lib/landing_slope/liblanding_slope.a src/lib/l1/libl1.a src/lib/pid/libpid.a src/lib/airspeed/libairspeed.a src/modules/sensors/data_validator/libdata_validator.a src/modules/sensors/vehicle_acceleration/libvehicle_acceleration.a src/modules/sensors/vehicle_angular_velocity/libvehicle_angular_velocity.a src/modules/sensors/vehicle_air_data/libvehicle_air_data.a src/modules/sensors/vehicle_imu/libvehicle_imu.a src/modules/sensors/sensor_calibration/libsensor_calibration.a src/lib/mag_compensation/libmag_compensation.a src/lib/drivers/accelerometer/libdrivers_accelerometer.a src/lib/drivers/barometer/libdrivers_barometer.a src/lib/drivers/gyroscope/libdrivers_gyroscope.a src/lib/drivers/device/libdrivers__device.a src/lib/drivers/magnetometer/libdrivers_magnetometer.a src/lib/conversion/libconversion.a src/lib/mathlib/libmathlib.a src/lib/battery/libbattery.a src/lib/ecl/geo/libecl_geo.a src/lib/mixer/libmixer.a src/lib/mixer/HelicopterMixer/libHelicopterMixer.a src/lib/mixer/MultirotorMixer/libMultirotorMixer.a src/lib/mixer/NullMixer/libNullMixer.a src/lib/mixer/SimpleMixer/libSimpleMixer.a src/lib/mixer/Mixer/libMixer.a src/lib/tunes/libtunes.a src/lib/version/libversion.a platforms/posix/src/px4/common/libpx4_layer.a src/lib/parameters/libparameters.a src/modules/uORB/libmodules__uORB.a platforms/common/px4_work_queue/libpx4_work_queue.a platforms/posix/src/px4/common/px4_daemon/libpx4_daemon.a platforms/common/libpx4_platform.a src/lib/systemlib/libsystemlib.a src/lib/cdev/libcdev.a platforms/posix/src/px4/common/libpx4_layer.a src/lib/parameters/libparameters.a src/modules/uORB/libmodules__uORB.a platforms/common/px4_work_queue/libpx4_work_queue.a platforms/posix/src/px4/common/px4_daemon/libpx4_daemon.a platforms/common/libpx4_platform.a src/lib/systemlib/libsystemlib.a src/lib/cdev/libcdev.a platforms/common/work_queue/libwork_queue.a boards/atlflight/excelsior/src/libdrivers_board.a src/lib/perf/libperf.a src/lib/parameters/tinybson/libtinybson.a msg/libuorb_msgs.a -lpthread -lm -lrt && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-uorb && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-batt_smbus && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-camera_trigger && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-ets_airspeed && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-ms4525_airspeed && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-ms5525_airspeed && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-sdp3x_airspeed && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-cm8jl65 && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-leddar_one && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-ll40ls && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-ll40ls_pwm && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-mappydot && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-mb12xx && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-pga460 && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-sf0x && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-sf1xx && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-srf02 && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-teraranger && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-tfmini && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-ulanding_radar && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-vl53l0x && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-gps && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-pwm_out_sim && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-qshell && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-rc_input && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-airspeed_selector && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-attitude_estimator_q && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-camera_feedback && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-commander && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-dataman && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-ekf2 && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-send_event && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-fw_att_control && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-fw_pos_control_l1 && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-land_detector && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-landing_target_estimator && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-local_position_estimator && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-logger && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-mavlink && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-mc_att_control && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-mc_hover_thrust_estimator && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-mc_pos_control && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-mc_rate_control && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-muorb && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-muorb_test && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-navigator && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-rc_update && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-rover_pos_control && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-sensors && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-simulator && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-battery_simulator && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-vmount && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-vtol_att_control && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-esc_calib && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-led_control && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-mixer && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-motor_ramp && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-motor_test && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-param && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-perf && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-pwm && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-sd_bench && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-shutdown && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-listener && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-tune_control && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-ver && cd /home/jake/code/jake/px4/Firmware/build/atlflight_excelsior_default/bin && /usr/bin/cmake -E create_symlink px4 px4-work_queue
../../src/lib/parameters/param.h:473: error: undefined reference to 'param_get'
../../src/lib/parameters/param.h:473: error: undefined reference to 'param_get'
../../src/lib/parameters/param.h:473: error: undefined reference to 'param_get'
../../src/lib/parameters/param.h:473: error: undefined reference to 'param_get'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
Makefile:224: recipe for target 'atlflight_excelsior' failed
make: *** [atlflight_excelsior] Error 1
```

Reply:
You need to add `#define PARAM_IMPLEMENTATION` in `parameters_shmem.cpp` before `#include "param.h"`

Reply:
That did the trick. Thank you!